### PR TITLE
[lit-html] Add keyed directive

### PR DESCRIPTION
**Pull request by justinfagnani**
Adds a `keyed(key, value)` directive that clears a part if the key changes. This is useful for ensuring stateful DOM is cleared and with animations that require adding or removing elements for fly-in / fly-out animations.

Fixes https://github.com/lit/lit/issues/2165

### CODE REVIEW

File: packages/lit-html/src/directives/keyed.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      setCommittedValue(part);
+      this.key = k;
+    }
+    return v;
+  }
+}
+
+/**
+ * Associates a renderable value with a unique key. When the key changes, the
+ * part is cleared forcing a new render, even if the value - such as a template
```
Comment (kevinpschaaf) [inline packages/lit-html/src/directives/keyed.ts:30]:
```suggestion
 * previous DOM is removed and disposed before rendering the next value, even if the value - such as a template
```

I think this clarifies the exact behavior better?

File: packages/lit-html/src/test/directives/keyed_test.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    const div = container.firstElementChild;
+    assert.equal(stripExpressionMarkers(container.innerHTML), '<div></div>');
+    assert.equal((div as any).foo, 1);
+
+    // Rerendering with same key should reuse the DOM
+    go(1);
+    const div2 = container.firstElementChild;
+    assert.equal(stripExpressionMarkers(container.innerHTML), '<div></div>');
+    assert.equal((div2 as any).foo, 1);
+    assert.strictEqual(div, div2, 'X');
```
Comment (kevinpschaaf) [inline packages/lit-html/src/test/directives/keyed_test.ts:34]: The `X` here is extraneous, right?

File: packages/lit-html/src/directives/keyed.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+  render(k: unknown, v: unknown) {
+    this.key = k;
+    return v;
+  }
+
+  override update(part: ChildPart, [k, v]: DirectiveParameters<this>) {
+    if (k !== this.key) {
+      // Clear the part before returning a value. This forces a re-render.
+      setCommittedValue(part);
```
Comment (kevinpschaaf) [inline packages/lit-html/src/directives/keyed.ts:21]: Man, had to look up the API for what no 2nd argument does. I remember now we defaulted it to a RESET_VALUE sigil that ensures any dirty check will fail. Maybe clarify in the comment that omitting the 2nd arg is what forces a re-render.

**Review by kevinpschaaf — APPROVED** [review]
Small nits.

File: packages/lit-html/src/test/directives/keyed_test.ts (hunk 2 above)
Comment (justinfagnani) [reply packages/lit-html/src/test/directives/keyed_test.ts:34]: Oh yeah, it's an assertion label for debugging. Will remove.

File: packages/lit-html/src/directives/keyed.ts (hunk 3 above)
Comment (justinfagnani) [reply packages/lit-html/src/directives/keyed.ts:21]: done

**Pull request merged**